File: backend/app/schemas/memory.py
from datetime import date, datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator


class MemoryCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")
    title: str = Field(min_length=1, max_length=80)
    content: str = Field(min_length=1, max_length=1200)
    kind: Literal["preference", "decision", "constraint", "note"] = "note"
    pinned: bool = False
    expires_on: date | None = None
    source_message_id: UUID | None = None

    @field_validator("title", "content")
    @classmethod
    def nonblank(cls, value):
        if not value.strip():
            raise ValueError("标题和内容不能为空")
        return value.strip()


class MemoryUpdate(MemoryCreate):
    revision: int = Field(ge=1)


class MemoryResponse(MemoryCreate):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    project_id: UUID | None
    revision: int
    created_at: datetime
    updated_at: datetime | None
    archived_at: datetime | None = None
    state: Literal["active", "expired", "archived"] = "active"
    index_status: Literal["pending", "ready", "failed"] = "pending"
    origin: Literal["manual", "automatic"] = "manual"
    source_quote: str | None = None


class MemoryPreferenceWrite(BaseModel):
    model_config = ConfigDict(extra="forbid")
    enabled: bool
    auto_extract: bool | None = None
    semantic_recall: bool | None = None
    extraction_model: str | None = Field(default=None, max_length=100)
    revision: int = Field(ge=0)


class MemoryPreview(BaseModel):
    model_config = ConfigDict(extra="forbid")
    query: str = Field(min_length=1, max_length=2000)


class MemoryUsageItem(BaseModel):
    id: str = Field(pattern=r"^[0-9a-fA-F-]{36}$")
    revision: int = Field(ge=1)


class MemoryUsage(BaseModel):
    retrieval_mode: Literal["hybrid", "keyword"] = "keyword"
    semantic_status: str | None = None
    status: Literal["used", "disabled", "no_match", "strict_knowledge"]
    items: list[MemoryUsageItem] = Field(max_length=6)
    omitted: int = Field(ge=0)
    estimated_tokens: int = Field(ge=0, le=1600)


class MemoryRevisionWrite(BaseModel):
    model_config = ConfigDict(extra="forbid")
    revision: int = Field(ge=1)


class MemoryProposalAccept(MemoryCreate):
    revision: int = Field(ge=1)


class ExtractedMemory(BaseModel):
    model_config = ConfigDict(extra="forbid")
    title: str = Field(min_length=1, max_length=80)
    content: str = Field(min_length=1, max_length=1200)
    kind: Literal["preference", "decision", "constraint", "note"]
    action: Literal["add", "update"]
    target: int | None = Field(default=None, ge=0, le=9)
    reason: str = Field(min_length=1, max_length=500)
    quote: str = Field(min_length=4, max_length=1500)
    expires_on: date | None = None
    expiry_quote: str | None = Field(default=None, max_length=300)


class ExtractionResult(BaseModel):
    memories: list[ExtractedMemory] = Field(default_factory=list, max_length=5)

File: frontend/src/lib/memory.ts
import { ApiError } from "@/lib/api-client";
import { currentProject } from "@/lib/project-scope";
import { useAuthStore } from "@/stores/auth-store";

export type MemoryKind = "preference" | "decision" | "constraint" | "note";
export type MemoryItem = {
  id: string;
  project_id: string | null;
  title: string;
  content: string;
  kind: MemoryKind;
  pinned: boolean;
  revision: number;
  source_message_id: string | null;
  created_at: string;
  updated_at: string | null;
};
export type MemoryList = { enabled: boolean; revision: number; items: MemoryItem[]; limit: number };
export type MemoryUsage = {
  status: "used" | "disabled" | "no_match" | "strict_knowledge";
  items: { id: string; revision: number }[];
  omitted: number;
  estimated_tokens: number;
};
export const memoryKey = () => [
  "project-memory",
  useAuthStore.getState().user?.id,
  currentProject()?.id ?? "default",
];
export const kindLabel = (kind: MemoryKind, zh: boolean) =>
  ({
    preference: ["回答偏好", "Preference"],
    decision: ["已确认决策", "Decision"],
    constraint: ["项目约束", "Constraint"],
    note: ["重要笔记", "Note"],
  })[kind][zh ? 0 : 1];
export function memoryError(error: unknown, zh: boolean) {
  if (error instanceof ApiError) {
    const data = error.data as { error?: { message?: string }; detail?: string } | undefined;
    if (zh && (data?.error?.message || typeof data?.detail === "string"))
      return (data?.error?.message || data?.detail) ?? (zh ? "操作失败" : "Request failed");
    if (error.status === 409)
      return zh
        ? "内容已更新或已存在，请刷新后重试。"
        : "Changed or duplicate content. Refresh and retry.";
  }
  return zh ? "操作失败，请稍后重试。" : "Request failed. Please retry.";
}
